rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,237 @@
+
+- Check the results of [the smoke tests](https://deploy.integration.publishing.service.gov.uk/job/Smokey/).
+- Look for any Icinga alerts related to the app.
+- Check the [deployment dashboard for the app](https://docs.publishing.service.gov.uk/manual/deployment-dashboards.html).
+- Generally make sure the change is tested.
+
+We often defer to automated tests to tell us when something is broken, but it's unclear whether this is really justified: there is no minimum standard for testing of GOV.UK apps.
+
+## Proposal
+
+The obvious solution to the backlog of deployments is to make them automatic. This has already begun, as part of [a trial of Continuous Deployment](https://docs.google.com/document/d/18y2fLzqbHyTliThxuiyxKznDw4TZmPiQZAk-3rkgpBg/edit) and [a subsequent extension to include 5 more apps](https://docs.google.com/document/d/1YhgjJjDRB57-IlgPNAAZzFTI3jIUucGJKokti_lQiRQ/edit).
+
+> **It is still possible to deploy an app manually.** The new pipeline is just an automation of the manual deployment steps. Rollback is unchanged, and freezing deployments is supported [[1](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb#L22)].
+
+However, we have been unable to make further progress due to the other two issues: we are concerned that enabling automatic deployments en masse would lead to an increase in defects reaching production.
+
+In order to proceed, we need to agree a clear and actionable set of criteria for enabling automatic deployments, so that we will only do so when we can show this is safe, as well as secure.
+
+### Security checks
+
+We need to ensure that only people with [production access](https://docs.publishing.service.gov.uk/manual/rules-for-getting-production-access.html) are able to deploy code to the live site. This is equivalent to being a member of the "GOV.UK Production" GitHub team. As part of the trial of CD, we restricted merge access on the affected repos to people in this team [[1](https://github.com/alphagov/govuk-saas-config/pull/49)] [[2](https://github.com/alphagov/govuk-saas-config/pull/51)] [[3](https://github.com/alphagov/govuk-saas-config/pull/55)]. We will [continue to use this approach](https://github.com/alphagov/govuk-saas-config/pull/52#issuecomment-651903746).
+
+As well as restricting who can merge, we also had to address a vulnerability with Integration Jenkins: someone without production access could manually trigger the CD pipeline by running the [`Deploy_App_Downstream` Jenkins job](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb). In order to prevent this kind of unauthorised deployment, the job has additional checks:
+
+- The tag to deploy must be the highest `release_123`-type tag.
+- The tagged commit must match the latest commit on the master branch.
```

Comment:
I'm a little unclear how this prevents an unauthorised deployments? Someone "unauthorised" could deploy the latest release? Is there still a risk that someone unauthorised could "flood" our deployment pipeline with a deployments?

Reply:
Good spot! 

I think this is a risk we will have to accept. Our security practices are just flawed: someone with malicious intent could use Integration in various ways to fool us into deploying dodgy code. This is another way, and we need to work with what we've got to restrict the attack surface - our Replatforming should do a far better job of this than we can with our current setup, where people have general SSH access to the whole Integration environment.

It's tempting to think about further restrictions. For example, we could try to detect if a release is already in production, but I don't think there's a secure way of doing that, and there are legitimate reasons for wanting to redeploy an existing release (albeit maybe not with the Deploy_App_Downstream job). One possible mitigation would be to throttle non-manual builds of Deploy_App in production to e.g. 10 an hour.

Otherwise, perhaps we should just call this out as a risk. How about this?

----

Someone without production access could still abuse the pipeline:

- By manually altering the job definition. In this scenario, we would permit an unauthorised deployment to the Staging environment, but no further: only people with production access can access Staging Jenkins [[1](https://github.com/alphagov/govuk-developer-docs/blame/736a167bf93b7338347c73cacbddfbdc68e72733/source/manual/rules-for-getting-production-access.html.md#L18-L19)], which also does the above checks.

- By repeatedly invoking the Deploy_App_Downstream job with the latest release, flooding the production environment with deployments. In this scenario, we would see a chain of [graceful restarts](https://github.com/alphagov/govuk-puppet/pull/10761), along with some loss of worker capacity.

Reply:
Yup - I agree there are bigger security holes and we should just call it out.

I'd be happy with calling it out as a risk like you said, except I'd rephrase it as these checks themselves don't prevented "unauthorised" deployments, they just mitigate against potentially malicious deployments that aren't the latest release.

I've added a suggestion below.

Reply:
Thanks! I've responded to your suggestion: https://github.com/alphagov/govuk-rfcs/pull/128#discussion_r514275852.

Continuing the discussion here, I think it's worth thinking a little about anything we could do to mitigate the risk you identified. My only suggestion so far is to apply some throttling to deployments in production, which seems like a fairly safe thing to do; it doesn't prevent abuse, but it limits the damage.

Do you have any suggestions?

Reply:
I've added [a work item](https://trello.com/c/cwsh79vA/200-rollout-of-continuous-deployment) to continue investigating this as part of the rollout of CD.

I'll resolve this now, although feel free to reopen if you'd like to continue discussion here.